Add `Column.shardingKeyIdx` for column definition for BanyanDB

wu-sheng opened the pull request:
<!--
    ⚠️ Please make sure to read this template first, pull requests that don't accord with this template
    maybe closed without notice.
    Texts surrounded by `<` and `>` are meant to be replaced by you, e.g. <framework name>, <issue number>.
    Put an `x` in the `[ ]` to mark the item as CHECKED. `[x]`
-->

<!-- ==== 🐛 Remove this line WHEN AND ONLY WHEN you're fixing a bug, follow the checklist 👇 ====
### Fix <bug description or the bug issue number or bug issue link>
- [ ] Add a unit test to verify that the fix works.
- [ ] Explain briefly why the bug exists and how to fix it.
     ==== 🐛 Remove this line WHEN AND ONLY WHEN you're fixing a bug, follow the checklist 👆 ==== -->

<!-- ==== 📈 Remove this line WHEN AND ONLY WHEN you're improving the performance, follow the checklist 👇 ====
### Improve the performance of <class or module or ...>
- [ ] Add a benchmark for the improvement, refer to [the existing ones](https://github.com/apache/skywalking/blob/master/apm-commons/apm-datacarrier/src/test/java/org/apache/skywalking/apm/commons/datacarrier/LinkedArrayBenchmark.java)
- [ ] The benchmark result.
```text
<Paste the benchmark results here>
```
- [ ] Links/URLs to the theory proof or discussion articles/blogs. <links/URLs here>
     ==== 📈 Remove this line WHEN AND ONLY WHEN you're improving the performance, follow the checklist 👆 ==== -->

<!-- ==== 🆕 Remove this line WHEN AND ONLY WHEN you're adding a new feature, follow the checklist 👇 ====
### <Feature description>
- [ ] If this is non-trivial feature, paste the links/URLs to the design doc.
- [ ] Update the documentation to include this new feature.
- [ ] Tests(including UT, IT, E2E) are added to verify the new feature.
- [ ] If it's UI related, attach the screenshots below.
     ==== 🆕 Remove this line WHEN AND ONLY WHEN you're adding a new feature, follow the checklist 👆 ==== -->

- [ ] If this pull request closes/resolves/fixes an existing issue, replace the issue number. Closes #<issue number>.
- [x] Update the [`CHANGES` log](https://github.com/apache/skywalking/blob/master/CHANGES.md).

* Add `Column.shardingKeyIdx` for column definition for BanyanDB.

```
Sharding key is used to group time series data per metric of one entity in one place (same sharding or same 
column for column-oriented database).
For example,
ServiceA's traffic gauge, service call per minute, includes following timestamp values, then it should be sharded by service ID
[ServiceA(encoded ID): 01-28 18:30 values-1, 01-28 18:31 values-2, 01-28 18:32 values-3, 01-28 18:32 values-4]

BanyanDB is the 1st storage implementation supporting this. It would make continuous time series metrics stored closely and compressed better.

NOTICE, this sharding concept is NOT just for splitting data into different database instances or physical files.
```

* Add `JD_PRESERVE_LINE_FEEDS=true` in official code style file. Then our comments don't have to use many blank lines to keep format.

wu-sheng commented on oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/storage/annotation/Column.java:
```diff
@@ -102,6 +101,28 @@
      */
     AnalyzerType analyzer() default AnalyzerType.OAP_ANALYZER;
 
+    /**
+     * Sharding key is used to group time series data per metric of one entity.
+     * For example,
+     * ServiceA's traffic gauge, service call per minute, includes following timestamp values, then it should be shard
+     * by service ID
+     * [ServiceA(encoded ID): 01-28 18:30 values-1, 01-28 18:31 values-2, 01-28 18:32 values-3, 01-28 18:32 values-4]
+     *
+     * BanyanDB is the 1st storage implementation supporting this. It would make continuous time series metrics stored
+     * closely and compressed better.
+     *
+     * 1. One entity at most has one sharding key
+     * 2. If no column is appointed for this, {@link org.apache.skywalking.oap.server.core.storage.StorageData#id}
+     * would
+     * be used by the storage implementation accordingly.
+     *
+     * NOTICE, this sharding concept is NOT for splitting data into different database instances or physical files.
+     *
+     * @return TRUE if this column could(not must) be used for sharding.
+     * @since 9.0.0
+     */
+    boolean shardingKey() default false;
```
@hanahmily @lujiajing1126 Please check this comment carefully. I want to be sure I understand this concept correctly.

wu-sheng commented:
![71743105-449D-41F8-B21D-C8B5C0D57DBA](https://user-images.githubusercontent.com/5441976/158975886-54fa226d-0a98-4d48-9a55-7c286ede1beb.png)

I have tested this for OAL generated metrics, it works.

lujiajing1126 commented:
I suppose we also need sharding keys for record, e.g. we use `is_error`, `service_id` and `service_instance_id` for `SegmentRecord`

lujiajing1126 commented on oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/storage/annotation/Column.java:
```diff
@@ -102,6 +101,28 @@
+     * BanyanDB is the 1st storage implementation supporting this. It would make continuous time series metrics stored
+     * closely and compressed better.
+     *
+     * 1. One entity at most has one sharding key
```
No. We support sharding with multiple keys now.

wu-sheng replied:
@hanahmily What do you prefer? I think you mentioned not to use multiple keys.

lujiajing1126 commented on oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/storage/annotation/Column.java:
```diff
@@ -102,6 +101,28 @@
+     * [ServiceA(encoded ID): 01-28 18:30 values-1, 01-28 18:31 values-2, 01-28 18:32 values-3, 01-28 18:32 values-4]
+     *
+     * BanyanDB is the 1st storage implementation supporting this. It would make continuous time series metrics stored
+     * closely and compressed better.
```
```suggestion
     * closely and compressed better. Besides, it can significantly alleviate high-cardinality issue for some indexed tags.
```

wu-sheng replied:
I think what you add is from database implementation, not has to be. Right?

lujiajing1126 commented on oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/storage/annotation/Column.java:
```diff
@@ -102,6 +101,28 @@
+     * For example,
+     * ServiceA's traffic gauge, service call per minute, includes following timestamp values, then it should be shard
+     * by service ID
+     * [ServiceA(encoded ID): 01-28 18:30 values-1, 01-28 18:31 values-2, 01-28 18:32 values-3, 01-28 18:32 values-4]
```
```suggestion
     * [ServiceA(encoded via hash function): 01-28 18:30 values-1, 01-28 18:31 values-2, 01-28 18:32 values-3, 01-28 18:32 values-4]
```

wu-sheng replied:
encoded means from OAP perspective. Via hash function is another thing from server side implementation. We should not document here.

wu-sheng commented:
> suppose we also need sharding keys for record, e.g. we use `is_error`, `service_id` and

@hanahmily any suggestion? We didn't discuss for traces and logs.

hanahmily commented:
> > suppose we also need sharding keys for record, e.g. we use `is_error`, `service_id` and
> 
> @hanahmily any suggestion? We didn't discuss for traces and logs.

After taking a look at segment and abstract log, I recommend the combination of entity pieces:

 * segment: `service_id`, `service_instance_id`, `is_error`
 * log: `service_id`,`service_instance_id`
 * other records: `id`, that's because these records don't need to be grouped by sharding key to get a better compression ratio.

wu-sheng commented:
@hanahmily @lujiajing1126 After confirmation that multiple columns should be supported, I changed `shardingKey` to `shardingKeyIdx`, which makes more sense as from both data insert/update or database sharding perspective, the sequence matters much. In the core level, I also added a checker to make sure the index is from zero and continuous.

Also, I added multiple-key sharding for trace(native and zipkin) and log(browser and general), using service ID, instance ID, and status.

FYI @mrproliu in the ebpf profiling, as the profiling record could be huge, please consider at least service ID as sharding key.

hanahmily reviewed (approved):
LGTM

lujiajing1126 reviewed (approved):
LGTM

kezhenxu94 commented on CHANGES.md:
```diff
@@ -112,14 +113,28 @@ Release Notes.
+
+BanyanDB is the 1st storage implementation supporting this. It would make continuous time series metrics stored closely and compressed better.
+
+NOTICE, this sharding concept is NOT just for splitting data into different database instances or physical files.
```
You keep using the term `shard` but explaining it's actually for grouping, what's the reason not to just use name like `Column.groupKeyIdx`?

lujiajing1126 replied:
From the prospetive of the acutual implementation, the value generated from the ordered combination of all sharkdingKeys here is used for sharding.

wu-sheng replied:
> Sharding key is used to group time-series data per metric of one entity in one place

> this sharding concept is NOT just for splitting data into different database instances or physical files

These 2 sentences are not conflicting. Group the key-equal data, and split the whole data set. If you feel confused, what do you prefer?
